src/caffeine/caffeine.c

```diff
@@ -33,130 +42,214 @@ void c_caffeinate(int argc, char *argv[])
   GASNET_SAFE(gex_Segment_Attach(&mysegment, myteam, segsz));
 }
 
-void c_decaffeinate(int exit_code)
+void caf_c_decaffeinate(int exit_code)
 {
   gasnet_exit(exit_code);
 }
 
-int c_this_image()
+int caf_c_this_image()
 {
   return gex_TM_QueryRank(myteam) + 1;
 }
 
-int c_num_images()
+int caf_c_num_images()
 {
   return gex_TM_QuerySize(myteam);
 }
 
-void c_sync_all()
+void caf_c_sync_all()
 {
   gasnet_barrier_notify(0,GASNET_BARRIERFLAG_ANONYMOUS);
   gasnet_barrier_wait(0,GASNET_BARRIERFLAG_ANONYMOUS);
 }
-void c_co_sum_no_result_image_int32(void* c_loc_a, size_t Nelem, int* stat)
+
+// GASNETT_INLINE(caf_c_co_reduce)
+void caf_c_co_reduce(
+  CFI_cdesc_t* a_desc, int result_image, int* stat, char* errmsg, int num_elements, gex_Coll_ReduceFn_t* user_op, void* client_data
+)
 {
-      gex_Event_t ev
-        = gex_Coll_ReduceToAllNB(myteam, c_loc_a, c_loc_a, GEX_DT_I32, sizeof(int32_t), Nelem, GEX_OP_ADD, NULL, NULL, 0);
-      gex_Event_Wait(ev);  
+  char* a_address = (char*) a_desc->base_addr;
+  size_t c_sizeof_a = a_desc->elem_len;
+  gex_Event_t ev;
 
-     if (stat != NULL) *stat = 0;
-}
+  switch (a_desc->type) {
+    case float_Complex_workaround:  num_elements *= 2; break;
+    case double_Complex_workaround: num_elements *= 2; break;
```

Comment:
I suspect these two lines are wrong, and that instead you want `c_sizeof_a *= 2`

ie. complex elements (the user-defined datatype) are twice as big, which is NOT equivalent to a reduction on twice as many float elements.

Reply:
Thanks, @bonachea.  All of the tests pass either way so I'm going with your suggested approach. I guess this could be a sign that the current test suite isn't stringent enough in that it doesn't test any reductions where this difference matters.

Reply:
I just pushed a commit with your suggested change.  I'll merge once all CI tests pass on the GitHub server.